ext/ScreenShot/ScreenShot.h

```diff
@@ -5,72 +5,186 @@
+				region.bufferRowLength = trueBufferRowLength;
+				region.bufferImageHeight = 0u;
+				region.imageOffset = { 0u, 0u, 0u };
+				region.imageExtent = image->getCreationParameters().extent;
+
+				video::IDriverMemoryBacked::SDriverMemoryRequirements memoryRequirements;
+				memoryRequirements.vulkanReqs.alignment = 64u;
+				memoryRequirements.vulkanReqs.memoryTypeBits = 0xffffffffu;
+				memoryRequirements.memoryHeapLocation = video::IDriverMemoryAllocation::ESMT_NOT_DEVICE_LOCAL;
+				memoryRequirements.mappingCapability = video::IDriverMemoryAllocation::EMCF_CAN_MAP_FOR_READ | video::IDriverMemoryAllocation::EMCF_COHERENT | video::IDriverMemoryAllocation::EMCF_CACHED;
+				memoryRequirements.vulkanReqs.size = image->getImageDataSizeInBytes();
+				auto destinationBuffer = driver->createGPUBufferOnDedMem(memoryRequirements);
+
+				auto fence = downloadImageMipLevel(driver, gpuImage.get(), destinationBuffer.get());
+				while (fence->waitCPU(1000ull, fence->canDeferredFlush()) == video::EDFR_TIMEOUT_EXPIRED) {}
 
-	auto buff = core::smart_refctd_ptr<video::IGPUBuffer>(driver->createDownStreamingGPUBufferOnDedMem((asset::getBytesPerPixel(outputFormatOverride)*texSize[0]*texSize[1]).getIntegerApprox()), core::dont_grab); // TODO
-	buff->getBoundMemory()->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ,{0u,buff->getSize()});
+				auto destinationBoundMemory = destinationBuffer->getBoundMemory();
+				destinationBoundMemory->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ, { 0u, memoryRequirements.vulkanReqs.size });
+				auto texelBuffer = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(memoryRequirements.vulkanReqs.size, destinationBoundMemory->getMappedPointer(), core::adopt_memory);
+				
+				auto mapPointerGetterFence = driver->placeFence(true);
+				while (mapPointerGetterFence->waitCPU(1000ull, mapPointerGetterFence->canDeferredFlush()) == video::EDFR_TIMEOUT_EXPIRED) {}
+				
+				destinationBoundMemory->unmapMemory();
 
-	auto fence = ext::ScreenShot::createScreenShot(driver, source, buff.get(), sourceMipLevel, 0ull,true,outputFormatOverride);
-	while (fence->waitCPU(1000ull, fence->canDeferredFlush()) == video::EDFR_TIMEOUT_EXPIRED) {}
-	ext::ScreenShot::writeBufferAsImageToFile(assetManager, _outFile, { texSize[0],texSize[1] }, outputFormatOverride, buff.get(), 0ull, flipY);
-}
+				image->setBufferAndRegions(std::move(texelBuffer), regions);
+				
+				asset::ICPUImageView::SCreationParams viewParams;
+				viewParams.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
+				viewParams.image = image;
+				viewParams.format = viewParams.image->getCreationParameters().format;
+				viewParams.viewType = asset::ICPUImageView::ET_2D;
+				viewParams.subresourceRange.baseArrayLayer = 0u;
+				viewParams.subresourceRange.layerCount = 1u;
+				viewParams.subresourceRange.baseMipLevel = 0u;
+				viewParams.subresourceRange.levelCount = 1u;
```

Comment:
I'd just make these match 100% with `image` and have the subresource encompass the whole `image` and be done with it.